README.md

```diff
@@ -1,16 +1,1261 @@
+    body: "This is the story of how one spec changed everything..."
+  }
+})
+
+const getEditors = crud.read("https://example.com/users/editors")
+
+const notify = msg.send("mailto:akiko@example.com", {
+  "to": await createDraft,
+  "subject": "Coffee",
+  "body": await getEditors
+})
+```
+
+While a Promise MAY be substituted for any field in a Closure, substituting the `do` field is NOT RECOMMENDED. The `do` field is critical in understanding what kind of action will be performed, and most schedulers will use the 
+
+After resolution, the Closure MUST be validated against the UCANs known to the Executor. A Promise resolved to a Closure that is not backed by a valid UCAN MUST NOT be executed, and SHOULD return an unauthorized error to the user.
+
+Promises MAY be used inside of a single Invocation's Closures, or across multiple Invocations, and MAY even be across multiple Invokers. As long as the pointer can be resolved, any invoked Task MAY be promised. This is sometimes referred to as ["promise pipelining"](http://erights.org/elib/distrib/pipeline.html).
+
+A Promise MUST resolve to a [Result](#8-result). If a particular branch's value is required to be unwrapped, the Result tag (`ok` or `err`) MAY be supplied.
+
+## 10.1 Fields
+
+``` ipldsch
+type Promise union {
+  | InvokedTaskPointer "ucan/ok"
+  | InvokedTaskPointer "ucan/err"
+  | InvokedTaskPointer "ucan/ptr"
+} representation keyed
+```
+
+If there are dependencies or ordering required, then you need a promise pipeline
+
+## 10.2 Pipelines
+
+Pipelining uses promises as inputs to determine the required dataflow graph. The following examples both express the following dataflow graph:
+
+### 10.2.1 Batched 
+
+![](./diagrams/batch-pipeline.svg)
+
+ ``` json
+{
+  "uiv": "0.1.0",
+  "nnc": "abcdef",
+  "prf": [
+    {"/": "bafkreie2cyfsaqv5jjy2gadr7mmupmearkvcg7llybfdd7b6fvzzmhazuy"},
+    {"/": "bafkreibbz5pksvfjyima4x4mduqpmvql2l4gh5afaj4ktmw6rwompxynx4"}
+  ],
+  "run": {
+    "update-dns" : {
+      "with": "dns:example.com?TYPE=TXT",
+      "do": "crud/update",
+      "inputs": { "value": "hello world"}
+    },
+    "notify-bob": {
+      "with": "mailto://alice@example.com",
+      "do": "msg/send",
+      "inputs": {
+        "to": "bob@example.com",
+        "subject": "DNSLink for example.com",
+        "body": {"ucan/ok": ["/", "update-dns"]}
+      }
+    },
+    "notify-carol": {
+      "with": "mailto://alice@example.com",
+      "do": "msg/send",
+      "inputs":{
+        "to": "carol@example.com",
+        "subject": "Hey Carol, DNSLink was updated!",
+        "body": {"ucan/ok": ["/", "update-dns"]}
+      }
+    },
+    "log-as-done": {
+      "with": "https://example.com/report",
+      "do": "crud/update",
+      "inputs": {
+        "payload": {
+          "from": "mailto://alice@exmaple.com",
+          "to": ["bob@exmaple.com", "carol@example.com"],
+          "event": "email-notification",
+        },
+        "_": [
+          {"ucan/ok": ["/", "notify-bob"]},
```

Comment:
I'm not big fan of this approach as it is not obvious if task handler is supposed to not get anything in the `_` input or if it should actually be provided the result of the ok branch ? I think I would much rather have something like `ucan/await` or `ucan/_` that resolves to `Null` that way you can stick those fields anywhere

Reply:
I had that in a previous version, but the feedback for a couple reviewers was that it would be better in the inputs because it's simpler to reason about. I honestly don't have strong feelings about this.

Reply:
Err, reading that again, we didn't have this part before:

> resolves to {} or Null that way you can stick those fields anywhere

We can put a field like `await: [Qm123, Qm456, Qm789]`, but now we have two mechanisms for describing temporal ordering.

We can namespace this as `ucan/_`, though it's common practice to call thrown-away fields `_`. The Task that returns a value can be anything, even if we don't need the value. This is here to add temporal constraints using the same mechanism that we already have (promises).

One of thinking about this strategy is a bit like [Haxl](https://engineering.fb.com/2014/06/10/web/open-sourcing-haxl-a-library-for-haskell/), Facebook's high performance data pipelining & query engine. It runs things in parallel unless there's a data dependency between them. Thanks to the [Church-Rosser Theorem](https://en.wikipedia.org/wiki/Church%E2%80%93Rosser_theorem), the runtime gains a huge amount of flexibility on all sugbraphs that have no direct dependency between them.

Reply:
Hmm yeah, the easiest thing may be to go back to adding an `after: [Promise]` field — it may be conceptually cleaner than altering/ignoring parts of the inputs

```js
{
  "with": "https://example.com/posts",
  "do": "crud/create",
  "inputs": {/* ... */},
  "await": [{"promise/ok": ["/", "some-earlier-task"]}]
}
```

AKA

```ipldsch
struct Task {
  with URI
  do String
  inputs Any
  await optional [Promise]
}
```

Reply:
The other criticism has been that the Task `meta` field gets abused for too many things. Here's an alternate that I _assume_ may be fairly common:

```haskell
type Task struct {
  with URI
  do String
  inputs Any
  after optional [Promise] -- Promise enables not running if not `ok`
  meta {String : Any} -- Extensibility, but mainly for humans
  resources {String : Any} -- Disk, memory, time, gas, etc
}
```